services/rfq/relayer/limiter/limiter.go

```diff
@@ -0,0 +1,162 @@
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Limiter is the interface for rate limiting RFQs.
+type Limiter interface {
+	// IsAllowed returns true if the request is allowed, false otherwise.
+	IsAllowed(ctx context.Context, request reldb.QuoteRequest) (bool, error)
+}
+
+type limiterImpl struct {
+	client     client.EVM
+	metrics    metrics.Handler
+	quoter     quoter.Quoter
+	cfg        relconfig.Config
+	tokenNames map[string]relconfig.TokenConfig
+}
+
+// NewRateLimiter creates a new Limiter.
+// TODO: implement the sliding window: queue up requests and process them in order if cumulative volume is above limit.
+func NewRateLimiter(
+	cfg relconfig.Config,
+	q quoter.Quoter,
+	client client.EVM,
+	metricHandler metrics.Handler,
+	tokens map[string]relconfig.TokenConfig,
+) Limiter {
+	return &limiterImpl{
+		// Limiter:    ratelimit.New(cfg.MaxRFQSize),
+		client:     client,
+		metrics:    metricHandler,
+		quoter:     q,
+		cfg:        cfg,
+		tokenNames: tokens,
+	}
+}
+
+// IsAllowed returns true if the request is allowed, false otherwise.
+func (l *limiterImpl) IsAllowed(ctx context.Context, request reldb.QuoteRequest) (_ bool, err error) {
+	ctx, span := l.metrics.Tracer().Start(ctx, "limiter.IsAllowed", trace.WithAttributes(
+		attribute.Int64("block_number", int64(request.BlockNumber)),
+		attribute.Int64("origin_token_decimals", int64(request.OriginTokenDecimals)),
+		attribute.Int64("dest_token_decimals", int64(request.DestTokenDecimals)),
+		attribute.String("transaction_id", fmt.Sprintf("%x", request.TransactionID)),
+		attribute.String("sender", request.Sender.Hex()),
+		attribute.String("origin_chain_id", fmt.Sprint(request.Transaction.OriginChainId)),
+		attribute.String("origin_token", request.Transaction.OriginToken.Hex()),
+		attribute.String("origin_amount", request.Transaction.OriginAmount.String()),
+		attribute.String("status", request.Status.String()),
+		attribute.String("origin_tx_hash", request.OriginTxHash.Hex()),
+		attribute.Int64("relay_nonce", int64(request.RelayNonce)),
+	))
+
+	defer func() {
+		metrics.EndSpanWithErr(span, err)
+	}()
+
+	// if enough confirmations, allow because reorgs are rare at this point
+	hasEnoughConfirmations, err := l.hasEnoughConfirmations(ctx, request)
+	if err != nil {
+		return false, fmt.Errorf("could not check confirmations: %w", err)
+	}
+	if hasEnoughConfirmations {
+		return true, nil
+	}
+
+	// if not enough confirmations, check volume
+	withinSize, err := l.withinSizeLimit(ctx, request)
+	if err != nil {
+		return false, fmt.Errorf("could not check volume limit: %w", err)
+	}
+
+	return withinSize, nil
+}
+
+func (l *limiterImpl) hasEnoughConfirmations(ctx context.Context, request reldb.QuoteRequest) (_ bool, err error) {
+	ctx, span := l.metrics.Tracer().Start(ctx, "limiter.hasEnoughConfirmations")
+	defer func() {
+		metrics.EndSpanWithErr(span, err)
+	}()
+
+	currentBlockNumber, err := l.client.BlockNumber(ctx)
```

Comment:
How often is this called? Honestly we could just pull this from listener to save some rpc calls. Depends how performance conciensce we want to be here

Add the very least I'd add a TODO about optimization and/or some kinda caching here

Reply:
Seperately, let's fetch requiredConfirmations first and fi ti's 0 return true. No reason to execute the rest of this routine otherwise (or even start the span)

Reply:
how does contract listener save RPC calls? 
i agree that client is overkill and probably is not the best choice for this though

Reply:
listener needs to track current blocknumber anyway and exposes it through 

https://pkg.go.dev/github.com/synapsecns/sanguine/ethergo@v1.1.0/listener#pkg-types

so you could pass in the listener instead as `type LatestBlockFetcher = interface {
    LatestBlock() int
}`